Build.props

```diff
@@ -0,0 +1,29 @@
+<Project>
```

Comment:
Is it possible to configure for a solution that we get warnings on build only when we explicitly want it (like with an MSBuild flag we talked about), otherwise just when we open a file?

The reason I'm asking is that this analysis makes compilation a lot longer. This is understandable of course, and we actually need it now in the beginning, when fixing all the issues first. However, after that, when we've fixed all the existing issues, during development we only need warnings for the developer when they open the given file. That way they'll see the issues but when they build locally they won't get the performance penalty. But in CI we'd still do a full build, so if they didn't care about fixing something it'll still break there.

Is this possible? So get full solution analyses on build in CI but only file-based analysis during local development?

Reply:
To disable code analysis locally when using Visual Studio you can update `Directory.Build.props` file in the repostory root as I mentioned here: https://github.com/Lombiq/Yevgeniys-plate/issues/1#issuecomment-663124267

Reply:
That's not what I'm talking about here. As I see it, analysis runs in two cases:
- When you open the file, just for that file.
- When you build the solution, for the whole solution.

The file-based one doesn't depend on the build because it also works when you open a freshly clones solution, before building it.

So my question is, can we configure this behavior in a way that during local development (by default) we'd only see warnings for the currently opened files and not run analysis on build (just in CI)?

Reply:
As far as I know there are only 2 options in Visual Studio which you can configure:

![image](https://user-images.githubusercontent.com/10154945/88645815-3a1be080-d0cd-11ea-8d28-272850d3e019.png)

![image](https://user-images.githubusercontent.com/10154945/88645862-456f0c00-d0cd-11ea-9527-52118bdbc250.png)

Reply:
Hmm, the "Background analysis scope" was already on "Open documents" for me. Doesn't seem to do much apparently.

The filter in the Error List window doesn't seem to do much for me, Build Only and IntelliSense Only shows exactly the same items.

So isn't there a way to disable analyzes during build?

Reply:
How about this? By default, the packages are removed on build (there are MSBuild steps running just before it) so it won't make the build slower. However, otherwise the packages are referenced so we get squiggly lines in the editor.

Then, you can disable this removal on build, i.e. enable code analysis on build, by either changing your solution's Directory.Build.props (for local development) or via a `dotnet build` command line parameter (for CI).

Reply:
I think analyzers should be enabled by default. During local development people can make a change to `Build.props` file to disable analyzers temporarily for faster builds.

Reply:
I agree that people should see analyzer warnings locally for what they work on. The issue is that running analysis for the whole solution, even if they only changed two lines, on every build, makes the build 2x slower. This is a hit on productivity. Hence why my suggestion of the above: If I'm not mistaken that technique would provide the best of both worlds, allowing warnings locally for open files while not making build slower, and at the same time allow full analysis on CI builds where the slowdown doesn't matter.

Reply:
I don't completely get it. You can configure it as you want.

Reply:
Maybe I'm not clear enough but I'd prefer you do this since it'll require digging into the depths of MSBuild.

So now these analyzer packages are added, and thus when you open a file you get analysis for just that file (this happens without any build, even if you open a freshly cloned solution). And when you do a build you get analysis for the whole solution.

Both of these are great but this makes the build much slower, which impacts productivity since during local development you need to iterate quickly and can't wait 5 minutes for every build. You also don't need to run analysis for the whole solution every time. So we need some solution for a developer to be able to see analyzer warnings when writing code but not be slowed down by slow builds. In CI though we always want a full thorough analysis and make the build fail when there are any issues that the developer didn't address.

The question is, how to achieve this? We don't want devs to completely disable analysis locally as they have to see issues for the code they work on, we need IDE support. Hence my idea:

- In CI everything happens as it happens now, so full analysis on build.
- Locally. the analyzer packages are referenced when editing but removed just when build happens. There are steps in MSBuild to hook into this event but I don't know from the top of my head. If I'm correct this will let analyzers run during editing but effectively disable them during build.

As I said, I'm happy to expand the scope and budget of our project for this.

Reply:
I found useful project property `<RunAnalyzersDuringBuild>false</RunAnalyzersDuringBuild>` and added it to `Build.props`. Look like it does what you wanted. During solution build, analyzers are not run. But Visual Studio shows analyzer warnings for open documents. And it doesn't affect `dotnet build`, which continue to find all analyzer warnings.

Please check it out, @Piedone.

Reply:
Here is this property on project properties window by the way:

![image](https://user-images.githubusercontent.com/10154945/90268431-b0f00200-de5f-11ea-9518-2dc2f9b1d6d4.png)

Reply:
Fantastic, that's just what we needed, thank you! Indeed a rebuild in VS now takes exactly the same amount of time than without the analyzers.

However, I noticed that the `dotnet build` output is not exactly the same. I've attached them [under the issue](https://github.com/Lombiq/Yevgeniys-plate/issues/1#issuecomment-674200033) because apparently that's not possible here. The "false" one is for `RunAnalyzersDuringBuild` = `false`.

I can't see any pattern apart from a bunch of warnings missing in the "false" file. I've built both with `dotnet build Finitive.sln --no-incremental > build.log` after each other ("false" first).

Reply:
Hmm, funnily enough, when executed in CI there are now more warnings then before.

And now if I run it again locally it's the same as with "true".

So it seems this was a red herring and works as expected!